src/lib/chawathe98_matcher.rs

```diff
@@ -46,35 +46,2262 @@
+    UPDATE,
+    /// Check the type of edge update which is insert
+    INSERT,
+    /// Check the type of edge update which is delete
+    DELETE,
+    /// Check the type of edge update which is move
+    MOVE,
+    /// Check the type of edge update which is copy
+    COPY,
+    /// Check the type of edge update which is glue
+    GLUE,
+    /// Check the type of edge update which is null
+    NULL,
+    /// Check the type of edge update which is Ok
+    OK,
+}
+#[derive(Debug, Clone)]
+/// The intial cost of each edge
+pub struct CostEdge {
+    /// Check the cost of edge update which is update
+    pub cost_update: usize,
+    /// Check the cost of edge update which is insert
+    pub cost_insert: usize,
+    /// Check the cost of edge update which is delete
+    pub cost_delete: usize,
+    /// Check the cost of edge update which is move
+    pub cost_move: usize,
+    /// Check the cost of edge update which is copy
+    pub cost_copy: usize,
+    /// Check the cost of edge update which is glue
+    pub cost_glue: usize,
+    /// Check the cost of edge update which is null
+    pub cost_null: usize,
+    /// Check the cost of edge update which is Ok
+    pub cost_ok: usize,
+}
+
+/// implemenation for edge cost
+
+impl CostEdge {
+    /// Create a new cost edge
+    /// This shows the cost for each edge defined by the user
+    pub fn new(
+        cost_update_1: usize,
+        cost_insert_1: usize,
+        cost_delete_1: usize,
+        cost_move_1: usize,
+        cost_copy_1: usize,
+        cost_glue_1: usize,
+        cost_null_1: usize,
+        cost_ok_1: usize,
+    ) -> CostEdge {
+        CostEdge {
+            cost_update: cost_update_1,
+            cost_insert: cost_insert_1,
+            cost_delete: cost_delete_1,
+            cost_move: cost_move_1,
+            cost_copy: cost_copy_1,
+            cost_glue: cost_glue_1,
+            cost_null: cost_null_1,
+            cost_ok: cost_ok_1,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+/// Edge construct which shows the edge between two nodes and their value
+pub struct Edge {
+    /// from node shows the node id from the From-Tree
+    pub from_node: NodeId<FromNodeId>,
+    /// to node shows the node id from To-Tree
+    pub to_node: NodeId<ToNodeId>,
+    /// The value of the edge
+    pub value: usize,
+    /// The edge type
+    pub edge_type: EdgeType,
+}
+
+/// Implementation for Edges
+impl Edge {
+    /// Create a new edge
+    pub fn new(
+        new_from_node: NodeId<FromNodeId>,
+        new_to_node: NodeId<ToNodeId>,
+        new_value: usize,
+        new_edge_type: EdgeType,
+    ) -> Edge {
+        Edge {
+            from_node: new_from_node,
+            to_node: new_to_node,
+            value: new_value,
+            edge_type: new_edge_type,
+        }
+    }
+}
+
+/// A store of mappings between nodes in different arenas.
+/// Direction is important.
+#[derive(Debug, Clone)]
+pub struct MappingStore2<T: Clone + Debug> {
```

Comment:
Can we call this something more descriptive than `MappingStore2`? Maybe `MappingStoreGraph` or `MappingStoreInducedGraph` or similar?